androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/DelegationService.java

```diff
@@ -14,19 +14,145 @@
+
 import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
 import androidx.browser.trusted.TokenStore;
+import androidx.browser.trusted.TrustedWebActivityCallbackRemote;
+import androidx.core.content.ContextCompat;
 
 /**
  * An extension of {@link androidx.browser.trusted.TrustedWebActivityService} that implements
  * {@link androidx.browser.trusted.TrustedWebActivityService#getTokenStore()} using a
  * {@link SharedPreferencesTokenStore}.
  */
 public class DelegationService extends androidx.browser.trusted.TrustedWebActivityService {
+    private static final String TAG = "TWADelegationService";
+
+    private static final String CHECK_LOCATION_PERMISSION_COMMAND_NAME = "checkAndroidLocationPermission";
+    private static final String LOCATION_PERMISSION_RESULT = "locationPermissionResult";
+
+    private static final String START_LOCATION_COMMAND_NAME = "startLocation";
+    private static final String STOP_LOCATION_COMMAND_NAME = "stopLocation";
+
+    private static final String EXTRA_RESULT_RECEIVER = "EXTRA_RESULT_RECEIVER";
+    private static final String EXTRA_PERMISSIONS = "EXTRA_PERMISSIONS";
+    private static final String EXTRA_GRANT_RESULTS = "EXTRA_GRANTED_RESULT";
+
+    private static final String mLocationPermission = Manifest.permission.ACCESS_FINE_LOCATION;
+
+    private boolean mIsProviderGmsCore;
 
     @NonNull
     @Override
     public TokenStore getTokenStore() {
         return new SharedPreferencesTokenStore(this);
     }
+
+    @Nullable
+    @Override
+    public Bundle onExtraCommand(
+            String commandName, Bundle args, TrustedWebActivityCallbackRemote callback) {
+        Log.d(TAG, commandName);
+        Bundle result = new Bundle();
+        switch (commandName) {
+            case CHECK_LOCATION_PERMISSION_COMMAND_NAME: {
+                if (callback == null) break;
+
+                result.putBoolean(commandName, true);
+                requestPermission(callback, mLocationPermission);
+                break;
+            }
+            case START_LOCATION_COMMAND_NAME: {
+                if (callback == null) return null;
+                if (!checkPermission(mLocationPermission)) return null;
+                startLocationProvider(callback, args.getBoolean("enableHighAccuracy"));
+
+                result.putBoolean(commandName, true);
+                break;
+            }
+            case STOP_LOCATION_COMMAND_NAME: {
+                stopLocationProvider();
+
+                result.putBoolean(commandName, true);
+                break;
+            }
+            default: {
+                result.putBoolean(commandName, false);
+            }
+        }
+        return result;
+    }
+
+    private boolean checkPermission(String permission) {
+        return ContextCompat.checkSelfPermission(this, permission) == PackageManager.PERMISSION_GRANTED;
+    }
+
+    private void requestPermission(TrustedWebActivityCallbackRemote callback, String permission) {
```

Comment:
Do you mind moving these methods into a separate class? We're going to be adding more stuff to the extraCommand logic (eg payments), so it would be nice if this class were pretty much just onExtraCommand and then the other logic was in different classes.

Reply:
I'm thinking that we can/should keep the basic method here: when these logic moving out of extraCommand, we still need the method in TWAService. I think it's nice to have the current structure align with what we meant to have in the future?